### PR TITLE
feat(blog website): Add a blog related to the allocation tracking work 

### PR DESCRIPTION
**Pull request by arshiyasolei**

### CODE REVIEW

File: website/content/en/blog/tracking-allocations.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+description: Gaining insight into Vector's per-component memory usage.
+authors: ["arshiyasolei", "tobz"]
+date: "2022-12-13"
+badges:
+  type: announcement
+  domains: ["monitoring"]
+tags: ["allocation-tracing", "tracking allocations"]
+---
+
+We are excited to announce that [Vector] now provides per-component memory usage metrics. To explore the feature, start a vector instance with the `allocation-tracing` option, launch `vector top` and monitor your components! This feature leverages a custom wrapping allocator combined with a separate reporting thread to collect our new metrics.
```
Comment (tobz) [inline website/content/en/blog/tracking-allocations.md:13]:
```suggestion
We are excited to announce that Vector now has support for exposing per-component memory usage metrics. This work begins to address an often-requested  feature from users who want to understand how Vector uses memory, and what parts of their configuration are responsible for high memory usage.

## Trying it out

To quickly try it out, you can pass `--allocation-tracing` when launching Vector which will enable the allocation tracing feature. This will emit new memory usage metrics as part of Vector's `internal_metrics` source output, as well as expose a new column in the `vector top` user interface to show the memory usage per component.
```

File: website/content/en/blog/tracking-allocations.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+We provide the following new metrics: `component_allocated_bytes`, `component_allocated_bytes_total`, and `component_deallocated_bytes_total`.
+
+- `component_allocated_bytes` captures the current net allocations/deallocations.
+- `component_allocated_bytes_total` shows the accumulated total allocations.
+- `component_deallocated_bytes_total` shows the accumulated total deallocations.
+
+## How it works
+
+From a high level point of view, we leverage our current component [tracing] infrastructure to track when a component "enters"/"exits" on a given thread. During each allocation/deallocation, we determine the responsible component via state stored in thread locals. Allocations that aren't associated with any components are tracked by a `root` component.
```
Comment (tobz) [inline website/content/en/blog/tracking-allocations.md:25]: This is probably a little too high level for most users.

I'd word this to something like:

```suggestion
Under the hood, Vector uses a custom memory allocator implementation which captures each time an allocation is made (or freed) and associates it with the currently-executing component. This works build upon some of the existing tracing functionality that Vector uses internal to provide structured logging, but has undergone a lot of work and effort around trying to optimize it for production usage.

As well, we also track allocations for the "root" component. The root component includes anything Vector allocates itself, regardless of whatever is specified in your configuration.
```

File: website/content/en/blog/tracking-allocations.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+## Notes
+
+This feature currently only supports unix based platforms. When enabled, there is approximately 20% overhead in throughput based on our benchmarks.
```
Comment (tobz) [inline website/content/en/blog/tracking-allocations.md:29]:
```suggestion
## Using allocation tracing in production

Currently, this feature is only supported on Linux builds of Vector. This may change in the future as we continue to improve allocation tracing.

Additionally, there is a small performance overhead imposed by allocation tracing when enabled. Vector runs a set of soak tests in CI in order to catch performance regressions during development. These soak tests emulate common Vector use cases, such as remapping events, or converting payloads from one event type to another, and so on.

In our development and testing of this feature, we've observed around a 20% reduction in throughput. This should be low enough to allow enabling it on a single Vector instance for debugging purposes, but care should be taken before enabling it across your entire fleet.
```

File: website/content/en/blog/tracking-allocations.md [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+## Next steps
+
+We currently do not provide support for determining shared memory usage between components. Adding support for shared ownership tracking provides further insights into the lifetimes of components, further easing the debugging process.
```
Comment (tobz) [inline website/content/en/blog/tracking-allocations.md:33]: We definitely need to reword this to explain the more intuitive reasoning. For example:

```
We currently only track memory usage in a way that shows which component is responsible for allocating memory, but not how that memory moves around Vector during normal operation.

For example, when a sink is batching events, and is configured to batch many events, you may observe that Vector is using a lot of memory. If you looked at the memory usage metrics, you would see most of it being attributed to components that either created the events (like the source) or touched the events (such as any transforms that modified the event) rather than the sink which is batching the events.

In this example, Vector will attribute that memory usage to the other components -- because that is where the initial allocations occurred -- but cannot yet properly account for the allocations moving to another component, where that new component is keeping the allocation alive (such as a sink batching events, as mentioned above) for some reason.

We plan on improving this in the future.
```

**Review by tobz — COMMENTED** [review]
Left some copyediting suggestions.

You don't necessarily need to apply them as-is, but I wanted to provide full examples so you can see where my mind was when thinking about how to describe/word this.

File: website/content/en/blog/tracking-allocations.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+tags: ["allocation-tracing", "tracking allocations"]
+---
+
+We are excited to announce that Vector now has support for exposing per-component memory usage metrics. This work addresses an often-requested feature from users who want to understand how Vector uses memory, and what parts of their configurations are responsible for high memory usage.
+
+## Trying it out
+
+![vector top with allocation tracing](/img/blog/vector-top-allocation-tracking.png)
+
+To explore the feature, start a Vector instance with the `allocation-tracing` option, launch `vector top` and monitor your components!
```
Comment (spencergilbert) [inline website/content/en/blog/tracking-allocations.md:19]: Minor point, but the `api` is disabled by default so `vector top` would fail without that config.

I'd also consider splitting this up to be something like:
---
To explore this feature, start a Vector instance with the `allocation-tracing` option:

```shell
vector --allocation-tracing -c /foo/config.yaml
```

Then, launch `vector top` and monitor your components!
---

Not personally familiar with the feature, I don't know if the option is a flag or in the config file...

File: website/content/en/blog/tracking-allocations.md [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+## What's new
+
+We provide the following new metrics: `component_allocated_bytes`, `component_allocated_bytes_total`, and `component_deallocated_bytes_total`.
+
+- `component_allocated_bytes` captures the current net allocations/deallocations.
```
Comment (spencergilbert) [inline website/content/en/blog/tracking-allocations.md:25]:
```suggestion
- `component_allocated_bytes` shows the current net allocations/deallocations.
```

Just for consistency with the other two descriptions.

File: website/content/en/blog/tracking-allocations.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+Under the hood, Vector uses a custom memory allocator implementation that captures each allocation/deallocation and associates it with the currently-executing component. This work builds upon Vector's existing tracing functionality. Allocations that aren't associated with any components are tracked by a `root` component.
+
+## Using allocation tracing in production
+
+Currently, this feature is only supported on Unix builds of Vector. This may change in the future as we continue to improve allocation tracing.
+
+Additionally, there is a performance overhead imposed by allocation tracing when enabled. Vector runs a set of soak tests in CI to catch performance regressions during development. These soak tests emulate common Vector use cases, such as remapping events, or converting payloads from one event type to another, and so on.
+
+In our development and testing of this feature, we've observed ~20% reduction in throughput. This overhead may be acceptable for debugging purposes, but care should be taken before enabling it across your entire fleet.
```
Comment (spencergilbert) [inline website/content/en/blog/tracking-allocations.md:42]: There's no reduction without the option enabled, right? I wonder if we need to say that explicitly just to cover ourselves and not worry someone 😄

File: website/content/en/blog/tracking-allocations.md [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+
+## Next steps
```
Comment (spencergilbert) [inline website/content/en/blog/tracking-allocations.md:44]: I don't know if there are existing issues, specifically for the mentioned enhancement, or just in general - but that could be nice to link. Also, a call-to-action around "let us know how it is/what is working well/what needs improvements/what are your dream features here" or what not, hopefully to drive some feedback from users to help direct work on this.

**Review by spencergilbert — COMMENTED** [review]
Couple of notes, otherwise a general nit about long-ish lines that could be line wrapped for easier readability.

File: website/content/en/blog/tracking-allocations.md [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+We currently do not provide support for determining shared memory usage between components. For example, when a sink is batching events, and is configured to batch many events, you may observe high Vector memory usage. If you looked at the memory usage metrics, you would see most of it being attributed to components that either created the events (such as a source) or processed the events (such as any transforms that modified the event) rather than the sink which is batching the events.
+
+In this example, Vector will attribute that memory usage to other components -- because that is where the initial allocations occurred -- but does not account for allocations moving to another component, where that new component is keeping the allocation alive (such as a sink batching events, as mentioned above).
+
+Adding support for shared ownership tracking provides further insights into the lifetimes of components, further easing the debugging process.
```
Comment (tobz) [inline website/content/en/blog/tracking-allocations.md:50]: This should be dropped, as the above text already describes the problem and that we plan to eventually tackle it.

File: website/content/en/blog/tracking-allocations.md [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+Additionally, there is a performance overhead imposed by allocation tracing when enabled. Vector runs a set of soak tests in CI to catch performance regressions during development. These soak tests emulate common Vector use cases, such as remapping events, or converting payloads from one event type to another, and so on.
+
+In our development and testing of this feature, we've observed ~20% reduction in throughput. This overhead may be acceptable for debugging purposes, but care should be taken before enabling it across your entire fleet.
+
+## Next steps
+
+We currently do not provide support for determining shared memory usage between components. For example, when a sink is batching events, and is configured to batch many events, you may observe high Vector memory usage. If you looked at the memory usage metrics, you would see most of it being attributed to components that either created the events (such as a source) or processed the events (such as any transforms that modified the event) rather than the sink which is batching the events.
```
Comment (tobz) [inline website/content/en/blog/tracking-allocations.md:46]: It feels weird to say that "we don't provide support", but maybe more than that, the example is talking about how we can't yet track how memory moves from one component to another in a logical fashion... whereas this wording is talking about "shared memory usage".

Nothing is being shared or double-counted, etc. I'll reiterate that this could stand to be reworded to focus on the notion that the allocation tracing currently doesn't fully track memory that moves from one component to another.

For reference, the first sentence of my original suggestion for rewriting this area:

> We currently only track memory usage in a way that shows which component is responsible for allocating memory, but not how that memory moves around Vector during normal operation.

File: website/content/en/blog/tracking-allocations.md (hunk 7 above)
Comment (tobz) [reply website/content/en/blog/tracking-allocations.md:42]: Correct, yeah. We should definitely bring that up that the feature has effectively no overhead when disabled.

**Review by tobz — CHANGES_REQUESTED** [review]
Overall, this is shaping up, but it still needs some work around the wording in order to properly communicate what it can and can't do, and make sure people understand what the metrics are indicating.

File: website/content/en/blog/tracking-allocations.md (hunk 5 above)
Comment (arshiyasolei) [reply website/content/en/blog/tracking-allocations.md]: Yeah the instructions do assume some familiarity with vector top

File: website/content/en/blog/tracking-allocations.md (hunk 10 above)
Comment (arshiyasolei) [reply website/content/en/blog/tracking-allocations.md]: Reworded it to mention ownership instead

**Review by spencergilbert — APPROVED** [review]
I'm good with this, assuming we don't feel the need to explain setting up `vector top` support as well.

**Review by tobz — APPROVED** [review]
